File: tests/browser-qa.mjs
import {createRequire} from 'node:module';
import {readFile,writeFile,mkdir,access} from 'node:fs/promises';
import path from 'node:path';
import {fileURLToPath} from 'node:url';
import {createPortfolioServer} from '../server.mjs';
const require=createRequire(import.meta.url);let chromium;
try{({chromium}=require('playwright'));}catch{({chromium}=require('../.work/browser/node_modules/playwright'));}
const root=fileURLToPath(new URL('..',import.meta.url)),out=path.join(root,'qa'),site=process.env.QA_ROOT||path.join(root,'dist');
await mkdir(out,{recursive:true});const data=JSON.parse(await readFile(path.join(site,'data/portfolio.json'),'utf8'));
const report={date:new Date().toISOString(),transport:'Real local HTTP navigation in installed Chrome; WebGL2 desktop and emulated mobile.',checks:[],errors:[],failedRequests:[]};
function check(name,pass,detail){report.checks.push({name,pass:!!pass,...(detail===undefined?{}:{detail})});console.log((pass?'PASS':'FAIL')+' '+name);}
const server=createPortfolioServer({root:site});await new Promise(resolve=>server.listen(0,'127.0.0.1',resolve));const origin='http://127.0.0.1:'+server.address().port;
const chrome=process.env.CHROME_PATH||(process.platform==='win32'?'C:/Program Files/Google/Chrome/Application/chrome.exe':undefined);
let browser;
try{
 browser=await chromium.launch({...(chrome?{executablePath:chrome}:{}),headless:true});
 const context=await browser.newContext({viewport:{width:1440,height:1000}});
 function observe(page){page.on('pageerror',e=>report.errors.push(e.message));page.on('response',r=>{if(r.url().startsWith(origin)&&r.status()>=400)report.failedRequests.push({url:r.url(),status:r.status()});});}
 const page=await context.newPage();observe(page);
 await page.goto(origin);check('Homepage has exactly 3 featured games',await page.locator('[data-project-card]').count()===3);
 await page.screenshot({path:path.join(out,'v6-home-desktop.png'),fullPage:true});
 await page.goto(origin+'/projects.html');check('Catalog contains 26 unique project cards',await page.locator('[data-project-card]').count()===26);
 await page.locator('#project-search').fill('thief');check('Thief search retains two distinct games',await page.locator('[data-project-card]:visible').count()===2);
 await page.locator('#reset-filters').click();await page.locator('#project-platform').selectOption('iOS');check('iOS filter selects the separate Apple app',await page.locator('[data-project-card]:visible').count()===1);
 for(const p of data.projects){
  await page.goto(origin+'/'+p.detailPage);
  const loaded=await page.locator('.project-detail img').evaluateAll(async images=>{await Promise.all(images.map(async i=>{i.loading='eager';try{await i.decode();}catch{}}));return images.every(i=>i.naturalWidth>0);});
  check(p.id+': complete local gallery, icon and estimate',loaded&&(await page.locator('[data-lightbox]').count())===p.media.screenshots.length&&(await page.locator('.detail-meta').innerText()).includes('ESTIMATED DEVELOPMENT TIME'));
 }
 await page.goto(origin+'/project-ballblast.html');await page.locator('[data-lightbox]').first().click();await page.keyboard.press('ArrowLeft');
 check('Lightbox wraps through all six Cannon Blast screenshots',(await page.locator('.lightbox-count').innerText())==='6 / 6');
 await page.keyboard.press('Escape');check('Lightbox restores keyboard focus',await page.locator('[data-lightbox]').first().evaluate(e=>e===document.activeElement));
 await page.goto(origin+'/3d-view.html?debug');await page.waitForFunction(()=>window.portfolioDebug?.ready,null,{timeout:45000});
 check('3D intro waits for a visitor click',await page.locator('#intro').isVisible());
 await page.locator('#start-button').click();await page.waitForTimeout(500);
 report.renderer=await page.evaluate(()=>portfolioDebug.game.engine.software?'Canvas':'WebGL2');check('Primary WebGL2 renderer starts',report.renderer==='WebGL2');
 check('Sound starts on the drive gesture',await page.evaluate(()=>portfolioDebug.game.audio.enabled&&portfolioDebug.game.audio.context.state==='running'));
 await page.evaluate(()=>portfolioDebug.travel('projects'));await page.waitForTimeout(1100);
 await page.screenshot({path:path.join(out,'v6-gallery-desktop.png')});
 check('Screen-space project labels are readable and clickable',await page.locator('.scene-label').count()>0);
 await page.locator('.scene-label').filter({hasText:'BloxStrike'}).click();check('Scene label opens its own project',await page.locator('#panel-title').innerText()==='BloxStrike');await page.keyboard.press('Escape');
 await page.evaluate(()=>portfolioDebug.travel('career-0'));await page.waitForTimeout(700);check('Career road shows first chapter',(await page.locator('#career-company').innerText()).includes('Gaminators'));
 await page.screenshot({path:path.join(out,'v6-career-desktop.png')});
 await page.evaluate(()=>portfolioDebug.travel('career-3'));await page.waitForTimeout(500);check('Career road reaches final chapter',(await page.locator('#career-company').innerText()).includes('OZI Technology'));
 await page.evaluate(()=>portfolioDebug.travel('career-1'));await page.waitForTimeout(500);check('Career road rewinds when travelling backwards',(await page.locator('#career-company').innerText()).includes('Mania'));
 await page.evaluate(()=>{const g=portfolioDebug.game;g.travel('blast-yard');const a=g.audio.context.createAnalyser();g.audio.master.connect(a);window.qaAudio=a;});await page.waitForTimeout(1100);
 const baseFov=await page.evaluate(()=>portfolioDebug.game.camera.fov);await page.screenshot({path:path.join(out,'v6-tnt-before.png')});
 await page.keyboard.down('KeyW');await page.waitForFunction(()=>portfolioDebug.game.world.destruction.stats.explosions>0,null,{timeout:6000});await page.keyboard.up('KeyW');await page.waitForTimeout(100);
 const blast=await page.evaluate(()=>{const g=portfolioDebug.game,a=new Float32Array(qaAudio.fftSize);qaAudio.getFloatTimeDomainData(a);return {fov:g.camera.fov,audioPeak:Math.max(...a.map(Math.abs)),particles:g.world.destruction.particles.filter(p=>p.active).length,simTime:g.simTime};});
 check('Keyboard-driven TNT collision produces debris',blast.particles>20);check('TNT camera zoom narrows the view',blast.fov<baseFov*.98);check('Explosion produces an audible signal',blast.audioPeak>.001,blast.audioPeak);
 await page.screenshot({path:path.join(out,'v6-tnt-explosion.png')});await page.waitForTimeout(300);const sim=await page.evaluate(()=>portfolioDebug.game.simTime);check('Blast slows the fixed-step simulation',sim-blast.simTime<.25,sim-blast.simTime);
 await page.waitForFunction(()=>portfolioDebug.game.world.destruction.stats.explosions===3,null,{timeout:6000});check('Nearby TNT crates create a three-crate chain reaction',true);
 await page.waitForTimeout(2100);check('Blast camera returns to its normal field of view',Math.abs((await page.evaluate(()=>portfolioDebug.game.camera.fov))-baseFov)<.01);
 await page.evaluate(()=>portfolioDebug.game.action('reset-props'));check('Reset restores every TNT, tree and pole',await page.evaluate(()=>portfolioDebug.game.world.destruction.props.every(p=>!p.broken&&p.body.enabled)));
 await page.evaluate(()=>{const g=portfolioDebug.game,p=g.world.destruction.props.find(p=>p.kind==='tree'&&p.position[0]===-8);g.vehicle.reset([-8,3,44]);g.blastAge=99;});await page.waitForTimeout(500);
 await page.keyboard.down('KeyW');await page.waitForFunction(()=>portfolioDebug.game.world.destruction.stats.trees>0,null,{timeout:6000});await page.keyboard.up('KeyW');check('Real keyboard driving breaks a tree',true);
 await page.screenshot({path:path.join(out,'v6-tree-break.png')});
 await page.evaluate(()=>portfolioDebug.game.setting('audio',false));check('Mute disables explosion and break audio',await page.evaluate(()=>!portfolioDebug.game.audio.enabled));
 await page.evaluate(()=>{const g=portfolioDebug.game;g.setting('reducedMotion',true);g.blastAge=.2;g.updateCamera(.016);});check('Reduced-motion preference suppresses camera zoom',Math.abs((await page.evaluate(()=>portfolioDebug.game.camera.fov))-baseFov)<.01);
 const mobile=await browser.newContext({viewport:{width:390,height:844},isMobile:true,hasTouch:true});const touch=await mobile.newPage();observe(touch);
 await touch.goto(origin);await touch.locator('.mobile-menu-toggle').tap();check('Mobile navigation opens with touch',await touch.locator('#site-nav').isVisible());check('Mobile homepage fits the viewport',await touch.evaluate(()=>document.documentElement.scrollWidth<=innerWidth));
 await touch.screenshot({path:path.join(out,'v6-home-mobile.png'),fullPage:true});
 await touch.goto(origin+'/3d-view.html?debug');await touch.waitForFunction(()=>window.portfolioDebug?.ready,null,{timeout:45000});await touch.locator('#start-button').tap();await touch.evaluate(()=>portfolioDebug.travel('projects'));await touch.waitForTimeout(1000);
 check('Mobile driving controls and TNT shortcut are visible',await touch.locator('#joystick').isVisible()&&await touch.locator('#explore-shortcuts [data-id="blast-yard"]').isVisible());
 check('Mobile shortcuts fit within the viewport',await touch.locator('#explore-shortcuts').evaluate(e=>{const b=e.getBoundingClientRect();return b.left>=0&&b.right<=innerWidth&&e.scrollWidth<=e.clientWidth;}));
 await touch.screenshot({path:path.join(out,'v6-gallery-mobile.png')});
 await touch.evaluate(()=>portfolioDebug.travel('career-0'));await touch.waitForTimeout(650);check('Mobile career story remains readable',await touch.locator('#career-story').isVisible());
 await touch.screenshot({path:path.join(out,'v6-career-mobile.png')});await touch.locator('#career-dismiss').tap();check('Career story can collapse on mobile',await touch.locator('#career-overlay').evaluate(e=>e.classList.contains('compact')));
 await mobile.close();await context.close();
 // Also exercise the fallback path with WebGL explicitly unavailable.
 const fallback=await browser.newContext({viewport:{width:800,height:600}});await fallback.addInitScript(()=>{const original=HTMLCanvasElement.prototype.getContext;HTMLCanvasElement.prototype.getContext=function(type,...args){return type==='webgl2'?null:original.call(this,type,...args);};});
 const compat=await fallback.newPage();observe(compat);await compat.goto(origin+'/3d-view.html?debug');await compat.waitForFunction(()=>window.portfolioDebug?.ready,null,{timeout:45000});await compat.locator('#start-button').click();check('Canvas compatibility renderer still starts',await compat.evaluate(()=>portfolioDebug.game.engine.software));await fallback.close();
 check('No uncaught browser exceptions',report.errors.length===0,report.errors);check('No missing local HTTP resources',report.failedRequests.length===0,report.failedRequests);
}catch(error){report.errors.push(error.stack);check('Browser QA completed',false,error.message);}
finally{await browser?.close();await new Promise(resolve=>server.close(resolve));report.pass=report.checks.every(c=>c.pass)&&!report.errors.length;await writeFile(path.join(out,'v6-browser-tests.json'),JSON.stringify(report,null,2));if(!report.pass)process.exitCode=1;console.log(`${report.checks.filter(c=>c.pass).length}/${report.checks.length} checks passed.`);}

File: tests/browser_qa.py
from pathlib import Path
from urllib.parse import urlparse,unquote
from playwright.sync_api import sync_playwright
import mimetypes,json,time
ROOT=Path(__file__).resolve().parents[1];OUT=ROOT/'qa';OUT.mkdir(exist_ok=True);report={'transport':'In-page local fixtures; browser URL navigation is administrator-blocked. HTML/JS/assets served via Playwright route fixtures, not a live deployment.','checks':[],'errors':[],'missing':[],'externalRequests':[]}
def check(name,value,detail=None):
 report['checks'].append({'name':name,'pass':bool(value),'detail':detail});print(('PASS' if value else 'FAIL'),name,detail or '',flush=True);(OUT/'browser-tests.json').write_text(json.dumps(report,indent=2))
with sync_playwright() as pw:
 browser=pw.chromium.launch(executable_path='/usr/bin/chromium',headless=True,args=['--no-sandbox'])
 def page_for(name,mobile=False):
  page=browser.new_page(viewport={'width':390,'height':844} if mobile else {'width':1440,'height':1000},is_mobile=mobile,has_touch=mobile,device_scale_factor=1)
  page.set_default_timeout(6000)
  page.on('pageerror',lambda e:report['errors'].append(str(e)))
  def route(r):
   u=urlparse(r.request.url)
   if u.hostname!='portfolio.test':report['externalRequests'].append(r.request.url);r.abort();return
   f=ROOT/(unquote(u.path).lstrip('/') or 'index.html')
   if f.is_file():r.fulfill(status=200,body=f.read_bytes(),headers={'Content-Type':mimetypes.guess_type(str(f))[0] or 'application/octet-stream','Access-Control-Allow-Origin':'*'})
   else:report['missing'].append(str(f));r.fulfill(status=404,body='Not found',headers={'Access-Control-Allow-Origin':'*'})
  page.route('**/*',route);page.evaluate('window.__PORTFOLIO_TEST__=true')
  html=(ROOT/name).read_text().replace('<head>','<head><base href="https://portfolio.test/">',1)
  page.set_content(html,wait_until='networkidle');page.wait_for_timeout(300)
  return page
 home=page_for('index.html');check('Home has exactly the first three projects',home.locator('[data-project-card]').count()==3)
 home.locator('img[src]').evaluate_all('(xs)=>xs.forEach(x=>x.loading="eager")');home.wait_for_function('Array.from(document.querySelectorAll("img[src]")).every(x=>x.complete&&x.naturalWidth>0)');check('Home contains local first-three images',home.locator('img[src]').evaluate_all('(xs)=>xs.every(x=>x.complete&&x.naturalWidth>0)'))
 check('Homepage has no horizontal overflow',home.evaluate('document.documentElement.scrollWidth<=innerWidth'))
 home.screenshot(path=str(OUT/'01-home-desktop.png'),full_page=True);home.close()
 projects=page_for('projects.html');check('Project page contains all 26 unique projects',projects.locator('[data-project-card]').count()==26)
 projects.locator('#project-search').fill('thief');check('Thief search keeps two distinct game packages',projects.locator('[data-project-card]:visible').count()==2)
 projects.locator('#reset-filters').click();projects.locator('#project-platform').select_option('iOS');check('Platform filter selects one iOS listing',projects.locator('[data-project-card]:visible').count()==1)
 projects.locator('#project-platform').select_option('');projects.locator('#project-category').select_option('Puzzle');check('Genre filter selects three puzzle games',projects.locator('[data-project-card]:visible').count()==3)
 projects.locator('#project-search').fill('not-a-project-123');check('No results has a visible helpful state',projects.locator('#empty-projects').is_visible());projects.close()
 for slug,expected in [('ballblast','5 working days'),('thief','About 1 month')]:
  page=page_for('project-'+slug+'.html');check(slug+' has the correct duration',expected in page.locator('.detail-meta').inner_text());check(slug+' no copied 5v5 description','5v5' not in page.locator('.project-detail').inner_text());page.close()
 detail=page_for('project-cubeshooter.html');detail.locator('[data-lightbox]').first.click();check('Detail gallery opens a real image viewer',detail.locator('.media-lightbox').is_visible());detail.keyboard.press('ArrowRight');check('Image viewer next key updates count',detail.locator('.lightbox-count').inner_text()=='2 / 4');detail.keyboard.press('Escape');check('Escape closes image viewer and restores focus',not detail.locator('.media-lightbox').is_visible() and detail.locator('[data-lightbox]').first.evaluate('(x)=>x===document.activeElement'));detail.close()
 profile=page_for('profile.html');check('Profile includes full contact information',all(v.lower() in profile.locator('main').inner_text().lower() for v in ['+923044744082','shehrazkhalid786@gmail.com','LinkedIn','Shehraz Khalid']));profile.close()
 world=page_for('3d-view.html');world.wait_for_function('window.portfolioDebug?.ready',timeout=30000)
 check('Onboarding is visible until user starts',world.locator('#intro').is_visible() and not world.evaluate('portfolioDebug.game.started'))
 world.locator('#start-button').click();world.wait_for_timeout(1500);check('Start reveals the driving world',not world.locator('#intro').is_visible() and world.locator('#hud').evaluate('(el)=>!el.hidden'))
 report['renderer']=world.evaluate('portfolioDebug.game.engine.software ? "Canvas compatibility renderer" : "WebGL2"')
 check('All scene project textures loaded',world.evaluate('portfolioDebug.game.world.projectExhibits.every(e=>e.screen.texture && e.icon.texture && e.screen.emissive===.32)'))
 world.evaluate("portfolioDebug.travel('about')");world.wait_for_timeout(1000)
 start=world.evaluate('portfolioDebug.snapshot().position');world.keyboard.down('KeyW');world.wait_for_timeout(650);world.keyboard.up('KeyW');end=world.evaluate('portfolioDebug.snapshot().position');check('Actual keyboard input moves the car',sum((a-b)**2 for a,b in zip(start,end))>0.05,{'before':start,'after':end})
 # Click the actual projected name sign, not a direct call to UI.show.
 world.evaluate("portfolioDebug.travel('about')");world.wait_for_timeout(1800)
 name=world.evaluate("(()=>{const g=portfolioDebug.game,t=g.world.hotspots.find(t=>t.id==='about'&&t.mesh.scale[0]>2);return {screen:g.engine.project(t.mesh.position),pick:g.experience.pick(...Object.values(g.engine.project(t.mesh.position)))?.id};})()")
 if name['screen']:
  world.mouse.click(name['screen']['x'],name['screen']['y']);world.wait_for_timeout(250)
 check('Clicking the 3D name board opens the basic profile',world.evaluate('portfolioDebug.game.ui.kind')=='about',name)
 if world.locator('#panel').is_visible():world.locator('#close-panel').click()
 world.evaluate("portfolioDebug.travel('projects')");world.wait_for_timeout(2300)
 world.screenshot(path=str(OUT/'04-project-exhibition.png'))
 target=world.evaluate("(()=>{const g=portfolioDebug.game,e=g.world.projectExhibits[0];const p=g.engine.project(e.screen.position);return {screen:p,pick:g.experience.pick(p.x,p.y)?.id};})()")
 world.mouse.click(target['screen']['x'],target['screen']['y']);world.wait_for_timeout(250)
 check('Clicking a physical project screenshot opens its full story',world.evaluate('portfolioDebug.game.ui.kind')=='project' and world.locator('#panel-title').inner_text()=='BloxStrike',target)
 if world.locator('#panel').is_visible():
  check('World project panel contains local gallery images',world.locator('#panel [data-lightbox]').count()==4)
  world.locator('#panel [data-lightbox]').first.click();check('World gallery enlarges screenshots',world.locator('.media-lightbox').is_visible());world.keyboard.press('Escape');check('Lightbox closes without closing the project panel',world.locator('#panel').is_visible());world.keyboard.press('Escape')
 # Single continuous physics drive; log chapters at actual changing vehicle positions.
 world.evaluate("portfolioDebug.travel('career-road')");world.wait_for_timeout(1400)
 world.screenshot(path=str(OUT/'05-career-start.png'))
 drive=world.evaluate("""(()=>{const g=portfolioDebug.game,seen=[];g.vehicle.controls={throttle:1,steer:0,boost:false,brake:false};for(let i=0;i<1800;i++){g.step();if(g.activeCareer>=0&&!seen.includes(g.activeCareer))seen.push(g.activeCareer);if(g.vehicle.body.p[0]>48)break;}g.vehicle.controls={throttle:0,steer:0,boost:false,brake:true};g.vehicle.body.v=[0,0,0];g.vehicle.body.w=[0,0,0];g.experience.update();g.camera.position=[g.vehicle.body.p[0]+6,20,-46];g.camera.target=[g.vehicle.body.p[0]+3,3,-60];return {seen,position:[...g.vehicle.body.p]};})()""")
 check('A continuous physics drive crosses all four career chapters',drive['seen']==[0,1,2,3],drive)
 world.wait_for_timeout(700);check('Career HUD reflects final company',world.locator('#career-company').inner_text()=='OZI Technology')
 world.screenshot(path=str(OUT/'06-career-ozi.png'))
 world.locator('#career-read').click();world.wait_for_timeout(150);check('Career read button opens full experience with selected chapter',world.locator('.career-entry.is-selected h3').inner_text()=='OZI Technology');world.keyboard.press('Escape')
 # Actual reverse gear over the same unobstructed lane.
 reverse=world.evaluate("""(()=>{const g=portfolioDebug.game,seen=[];g.vehicle.reset([48,3,-58],-Math.PI/2);g.vehicle.controls={throttle:-1,steer:0,boost:false,brake:false};for(let i=0;i<2000;i++){g.step();if(g.activeCareer>=0&&!seen.includes(g.activeCareer))seen.push(g.activeCareer);if(g.vehicle.body.p[0]<-42)break;}g.vehicle.controls={throttle:0,steer:0,boost:false,brake:true};g.vehicle.body.v=[0,0,0];g.vehicle.body.w=[0,0,0];g.experience.update();g.camera.position=[g.vehicle.body.p[0]+6,20,-46];g.camera.target=[g.vehicle.body.p[0]+3,3,-60];return {seen,position:[...g.vehicle.body.p]};})()""")
 check('Reversing through the lane rewinds all career chapters',reverse['seen']==[3,2,1,0],reverse)
 world.wait_for_timeout(500);check('Reverse HUD returns to first employer','Gaminators' in world.locator('#career-company').inner_text())
 world.evaluate("portfolioDebug.travel('skills')");world.wait_for_timeout(200);check('Leaving career lane hides career overlay',not world.locator('#career-overlay').is_visible())
 world.evaluate("portfolioDebug.game.ui.show('map')");check('Map contains a working career-road destination',world.locator('#panel .career-map-stop').is_visible());world.locator('#panel .career-map-stop').click();world.wait_for_timeout(200);check('Map travel enters career lane',world.evaluate('portfolioDebug.game.activeCareer')==0)
 world.close()
 mobile=page_for('index.html',True);check('Mobile homepage has no horizontal overflow',mobile.evaluate('document.documentElement.scrollWidth<=innerWidth'));mobile.locator('.mobile-menu-toggle').click();check('Mobile navigation opens',mobile.locator('.site-nav').is_visible());mobile.locator('.mobile-menu-toggle').click();mobile.screenshot(path=str(OUT/'07-home-mobile.png'),full_page=True);mobile.close()
 mobileproj=page_for('project-thief.html',True);check('Mobile project details fit width',mobileproj.evaluate('document.documentElement.scrollWidth<=innerWidth'));mobileproj.screenshot(path=str(OUT/'08-thief-mobile.png'),full_page=True);mobileproj.close()
 mobileworld=page_for('3d-view.html',True);mobileworld.wait_for_function('window.portfolioDebug?.ready',timeout=30000);mobileworld.locator('#start-button').click();mobileworld.wait_for_timeout(700);mobileworld.evaluate("portfolioDebug.travel('projects')");mobileworld.wait_for_timeout(1300)
 check('Mobile touch controls appear',mobileworld.locator('#joystick').is_visible())
 # Center icon is tappable in portrait; a separate test checks screens on desktop.
 xy=mobileworld.evaluate("(()=>{const g=portfolioDebug.game,e=g.world.projectExhibits[1];const p=g.engine.project(e.icon.position);return {x:p.x,y:p.y,pick:g.experience.pick(p.x,p.y)?.id};})()")
 mobileworld.screenshot(path=str(OUT/'09-projects-world-mobile.png'))
 mobileworld.touchscreen.tap(xy['x'],xy['y']);mobileworld.wait_for_timeout(300)
 check('Mobile tap on physical project icon opens Cannon Blast details',mobileworld.evaluate('portfolioDebug.game.ui.kind')=='project' and 'Cannon' in mobileworld.locator('#panel-title').inner_text(),xy)
 if mobileworld.locator('#panel').is_visible():mobileworld.locator('#close-panel').click()
 mobileworld.evaluate("portfolioDebug.travel('career-road')");mobileworld.wait_for_timeout(1300)
 check('Mobile career narrative visible',mobileworld.locator('#career-overlay').is_visible())
 mobileworld.screenshot(path=str(OUT/'10-career-mobile.png'));mobileworld.close()
 browser.close()
check('No uncaught browser exceptions',not report['errors'],report['errors'])
check('No missing local runtime assets',not report['missing'],report['missing'])
check('No external runtime font/image/CDN requests',not report['externalRequests'],report['externalRequests'])
report['passed']=sum(c['pass'] for c in report['checks']);report['total']=len(report['checks']);(OUT/'browser-tests.json').write_text(json.dumps(report,indent=2))
print('RESULT',report['passed'],'/',report['total'])
